File: playground/multi_workspace_roundtrip.py
#!/usr/bin/env python3
"""Cross-store (.cross.h5) round-trip suite (M2.4/M2.7).

Drives the fts_cross_roundtrip C++ CLI and validates the file structure with
h5py (Python can reach the FILE level, never AppState):

  1. create            -> empty archive v2 (no root @format)
  2. add 2 sources     -> structure valid; workspace.json byte-identical
                         (view-state fidelity, audit §2.3 item 6)
  3. load a source     -> CLI dump matches the h5py-read content
  4. save-back         -> source group rewritten, structure still valid
  5. remove            -> group + manifest entry gone
  6. atomicity         -> SIGKILL mid-save (--slow-save): the original archive
                         is intact and opens; a stale .tmp is tolerated and
                         cleaned by the next mutation
  7. version gate      -> manifest version 3 is refused

Usage:
    python3 playground/multi_workspace_roundtrip.py [--binary PATH]

Requires h5py. Outputs land in playground/outputs/multi_workspace_roundtrip/.
"""

import argparse
import json
import os
import shutil
import signal
import subprocess
import sys
import time
from pathlib import Path

HERE = Path(__file__).resolve().parent
REPO_ROOT = HERE.parent
OUT = REPO_ROOT / "playground" / "outputs" / "multi_workspace_roundtrip"
BIN_DIR_CANDIDATES = (
    REPO_ROOT / "build" / "linux-release",
    REPO_ROOT / "build" / "linux-debug",
    REPO_ROOT / "build",
)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--binary", default=None, help="fts_cross_roundtrip binary")
    args = ap.parse_args()

    binary = Path(args.binary) if args.binary else None
    if binary is None:
        for cand in BIN_DIR_CANDIDATES:
            cand = cand / "fts_cross_roundtrip"
            if cand.exists():
                binary = cand
                break
    if binary is None:
        print("fts_cross_roundtrip binary not found; build first")
        return 1

    shutil.rmtree(OUT, ignore_errors=True)
    OUT.mkdir(parents=True)
    work = OUT / "work"
    work.mkdir()

    def run(*cmd):
        res = subprocess.run([str(binary), *map(str, cmd)], capture_output=True, text=True)
        if res.returncode != 0:
            print(f"FAIL: {' '.join(cmd)} -> {res.returncode}\n{res.stderr}")
            sys.exit(1)
        return res.stdout.strip()

    def h5open(path):
        import h5py
        return h5py.File(path, "r")

    # 1. create
    cross = work / "cross.h5"
    run("create", cross)
    with h5open(cross) as f:
        assert list(f.keys()) == ["archive.json"], f.keys()
        assert "format" not in f.attrs, "cross root must not carry @format"
        manifest = json.loads(f["archive.json"][()])
        assert manifest["version"] == 2
        assert manifest["sources"] == []
    print("1. create: OK")

    # 2. add two sources (the second add is a dedupe -> srcA_2)
    src1 = work / "src_a.h5"
    src2 = work / "src_b.h5"
    make_source_workspace(src1, "source A", {"zoom": [1, 2, 3]})
    make_source_workspace(src2, "source B", {"zoom": [9, 8]})
    assert run("add", cross, src1) == "src_a"
    assert run("add", cross, src2) == "src_b"
    assert run("add", cross, src1) == "src_a_2"
    with h5open(cross) as f:
        manifest = json.loads(f["archive.json"][()])
        assert [s["id"] for s in manifest["sources"]] == ["src_a", "src_b", "src_a_2"]
        for sid in ("src_a", "src_b"):
            g = f[f"sources/{sid}"]
            assert g.attrs["format"] == "unified-spectral-data-container"
            summary = json.loads(g.attrs["summary"])
            assert summary["memberCount"] == 2
            # Fidelity: the view state round-trips exactly (semantic identity).
            # The C++ model normalizes JSON (nlohmann: compact, sorted keys) —
            # app-written workspace.json round-trips byte-identically; the
            # python-written fixture is compared in canonical form.
            orig = json.loads(h5py_read_workspace_json(src1 if sid == "src_a" else src2))
            embedded = json.loads(g["workspace.json"][()])
            assert embedded == orig, f"workspace.json drift in {sid}"
    print("2. add + dedupe + fidelity: OK")

    # 3. load a source: CLI dump must match the h5py-read content
    dump = run("load", cross, "src_a")
    assert "uncorrected=2 corrected=0 spectra=0" in dump, dump
    assert "comment=source A" in dump, dump
    print("3. load: OK")

    # 4. save-back (whole-source rewrite) with a MODIFIED comment
    tweaked = work / "src_a_tweaked.h5"
    make_source_workspace(tweaked, "source A EDITED", {"zoom": [1, 2, 3]})
    run("save-source", cross, "src_a", tweaked)
    dump = run("load", cross, "src_a")
    assert "comment=source A EDITED" in dump, dump
    with h5open(cross) as f:
        assert "src_b" in f["sources"], "save-back must not touch other sources"
    print("4. save-back: OK")

    # 5. remove
    run("remove", cross, "src_a_2")
    with h5open(cross) as f:
        manifest = json.loads(f["archive.json"][()])
        assert [s["id"] for s in manifest["sources"]] == ["src_a", "src_b"]
        assert "src_a_2" not in f["sources"]
    print("5. remove: OK")

    # 6. atomicity: kill during a slow save
    before = cross.read_bytes()
    proc = subprocess.Popen(
        [str(binary), "add", str(cross), str(src2), "--slow-save"],
        stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    time.sleep(1.0)                       # inside the 2 s slow-save window
    proc.send_signal(signal.SIGKILL)
    proc.wait()
    assert cross.read_bytes() == before, "archive modified by killed save"
    with h5open(cross) as f:              # still opens + lists
        assert len(json.loads(f["archive.json"][()])["sources"]) == 2
    print("6. atomicity: OK")

    # 7. version gate
    bad = work / "bad_version.h5"
    run("create", bad)
    import h5py
    with h5py.File(bad, "r+") as f:
        f["archive.json"][()] = json.dumps({"version": 3, "sources": []})
    res = subprocess.run([str(binary), "list", str(bad)], capture_output=True, text=True)
    assert res.returncode != 0
    assert "unsupported archive version 3" in res.stderr, res.stderr
    print("7. version gate: OK")

    print(f"multi_workspace_roundtrip: ALL OK (workdir {work})")
    return 0


def make_source_workspace(path, comment, view_json):
    import h5py
    import numpy as np
    with h5py.File(path, "w") as f:
        f.attrs["format"] = "unified-spectral-data-container"
        f.attrs["created"] = "2026-08-01T00:00:00Z"
        f.create_dataset("measurement_comment.txt", data=comment)
        f.create_dataset("measurement_config.json",
                         data=json.dumps({"instrument": {"model": "roundtrip"}}))
        f.create_dataset("tags", data="ftir, test")
        f.create_dataset("workspace.json",
                         data=json.dumps({"applications": {"FTS Data Explorer": view_json}}))
        g = f.create_group("igm_uncorrected_x")
        g.attrs["schema"] = "interferogram"
        g.create_dataset("record_0", data=np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32))
        g["record_0"].attrs["kind"] = "original"
        g.create_dataset("record_1", data=np.array([[5.0, 6.0], [7.0, 8.0]], dtype=np.float32))
        g["record_1"].attrs["kind"] = "original"


def h5py_read_workspace_json(path):
    import h5py
    with h5py.File(path, "r") as f:
        return f["workspace.json"][()]


if __name__ == "__main__":
    sys.exit(main())

File: session/multi_workspace_roundtrip.cpp
// fts_multi_workspace_roundtrip — CLI driver for the multi-workspace .h5 store
// (M2.4), driven by playground/multi_workspace_roundtrip.py (h5py can read the
// file structure but cannot call the C++ store API).
//
// Usage:
//   fts_multi_workspace_roundtrip create <path>
//   fts_multi_workspace_roundtrip add <multi-workspace.h5> <src.h5> [--slow-save]
//   fts_multi_workspace_roundtrip remove <multi-workspace.h5> <id>
//   fts_multi_workspace_roundtrip list <multi-workspace.h5>
//   fts_multi_workspace_roundtrip load <multi-workspace.h5> <id>
//   fts_multi_workspace_roundtrip sniff <path>
//   fts_multi_workspace_roundtrip save-source <multi-workspace.h5> <id> <src.h5>
// Exit 0 = success, 1 = failure (message on stderr).

#include <cinttypes>
#include <cstdio>
#include <cstring>
#include <string>

#include "app_state.h"
#include "multi_workspace_store.h"
#include "hdf/h5_store.h"

namespace {

int fail(const std::string& msg) {
    std::fprintf(stderr, "%s\n", msg.c_str());
    return 1;
}

int cmdCreate(const std::string& path) {
    std::string err;
    if (!multiWorkspaceCreate(path, err)) return fail(err);
    std::printf("created %s\n", path.c_str());
    return 0;
}

int cmdAdd(const std::vector<std::string>& args) {
    bool slow = false;
    std::string multiWorkspacePath, srcPath;
    for (const auto& a : args) {
        if (a == "--slow-save") slow = true;
        else if (multiWorkspacePath.empty()) multiWorkspacePath = a;
        else if (srcPath.empty()) srcPath = a;
    }
    if (multiWorkspacePath.empty() || srcPath.empty())
        return fail("usage: add <multi-workspace.h5> <src.h5> [--slow-save]");
    std::string err, newId;
    if (!multiWorkspaceAddSource(multiWorkspacePath, srcPath, newId, err, slow)) return fail(err);
    std::printf("%s\n", newId.c_str());
    if (slow) std::printf("slow-save: 2s window for the kill test\n");
    return 0;
}

int cmdRemove(const std::string& multiWorkspacePath, const std::string& id) {
    std::string err;
    if (!multiWorkspaceRemoveSource(multiWorkspacePath, id, err)) return fail(err);
    std::printf("removed %s\n", id.c_str());
    return 0;
}

int cmdList(const std::string& multiWorkspacePath) {
    SessionTabState st;
    std::string err;
    if (!multiWorkspaceLoadInto(st, multiWorkspacePath, err)) return fail(err);
    for (const auto& src : st.sources)
        std::printf("%s\t%s\t%zu\t%" PRIu64 "\n", src.id.c_str(), src.name.c_str(),
                    src.memberCount, src.sizeBytes);
    return 0;
}

int cmdLoad(const std::string& multiWorkspacePath, const std::string& id) {
    std::string err;
    Workspace ws = multiWorkspaceLoadSource(multiWorkspacePath, id, err);
    if (!err.empty()) return fail(err);
    std::printf("format=%s\n", ws.format.c_str());
    std::printf("comment=%s\n", ws.measurementComment.c_str());
    std::printf("tags=%s\n", ws.tags.c_str());
    std::printf("uncorrected=%zu corrected=%zu spectra=%zu\n",
                ws.uncorrectedIfg.members.size(), ws.correctedIfg.members.size(),
                ws.spectra.members.size());
    std::printf("workspaceJson=%s\n", ws.workspaceJson.dump().c_str());
    return 0;
}

int cmdSniff(const std::string& path) {
    std::printf(isMultiWorkspaceFile(path) ? "multi-workspace\n" : "workspace\n");
    return 0;
}

int cmdSaveSource(const std::string& multiWorkspacePath, const std::string& id,
                  const std::string& srcPath) {
    Workspace ws;
    try {
        ws = H5Store::load(srcPath);
    } catch (const std::exception& e) {
        return fail(std::string("load source failed: ") + e.what());
    }
    std::string err;
    try {
        multiWorkspaceSaveSource(multiWorkspacePath, id, ws, err);
    } catch (const std::exception& e) {
        return fail(e.what());
    }
    if (!err.empty()) return fail(err);
    std::printf("saved %s\n", id.c_str());
    return 0;
}

}  // namespace

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::fprintf(stderr, "usage: fts_multi_workspace_roundtrip create|add|remove|list|load|sniff|save-source ...\n");
        return 1;
    }
    const std::string mode = argv[1];
    if (mode == "create" && argc == 3) return cmdCreate(argv[2]);
    if (mode == "add") {
        std::vector<std::string> args(argv + 2, argv + argc);
        return cmdAdd(args);
    }
    if (mode == "remove" && argc == 4) return cmdRemove(argv[2], argv[3]);
    if (mode == "list" && argc == 3) return cmdList(argv[2]);
    if (mode == "load" && argc == 4) return cmdLoad(argv[2], argv[3]);
    if (mode == "sniff" && argc == 3) return cmdSniff(argv[2]);
    if (mode == "save-source" && argc == 5) return cmdSaveSource(argv[2], argv[3], argv[4]);
    std::fprintf(stderr, "bad arguments\n");
    return 1;
}
